### PR TITLE
Game end2

### PR DESCRIPTION
**Pull request by character331**

### CODE REVIEW

File: app/game/game.js [hunk 1]
```diff
@@ -85,7 +86,25 @@ const GAME = (humanUsername, aiUsername, gameId) => {
...
+                CLIENTS_HANDLER.updatePlayerStats(HUMAN_USERNAME, 'You lost');
+                CLIENTS_HANDLER.updatePlayerStats(AI_USERNAME, 'You lost');
+                removeGame(gameId);
+            } else {
+                sendDataToPlayer(gameId, HUMAN_USERNAME, GameEndData('You lost'));
+                sendDataToPlayer(gameId, AI_USERNAME, GameEndData('You lost'));
+                console.log("At least one user is not loggedin");
+                removeGame(gameId);
+            }
+        }
```
Comment (ColinJPage) [inline app/game/game.js:106]: The `tick()` function is not a great place for this. When the time runs out, we could call an `onTimeOut()` function. If the ship explodes, we call `onShipExplode()`. After these functions notify the clients and update stats, both can call `endGame()` which removes the Game from GAMES and does any other needed clean-up.

File: app/game/game.js [hunk 2]
```diff
@@ -258,11 +277,18 @@ const lookUpRole = (gameId, username) => {
...
     }
 }
 
+const removeGame = (gameId) => {
+    delete GAMES[gameId];
+    console.log("game has been removed", GAMES);
+}
+
 setInterval(tickGames, GAME_TICK_DELAY_MS);
+setTimeout(() => removeGame(1), 5000);
```
Comment (ColinJPage) [inline app/game/game.js:292]: This shouldn't be merged, right?

File: app/public/board.js [hunk 3]
```diff
@@ -5,6 +5,31 @@ const BOARD = ((rowCount, columnCount) => {
...
+        text-align: center;
+        z-index: 2;
+        display: none; 
+        }
+
+        #message-text {
+        font-size: 24px;
+        margin: 0;
+        }
+     */
```
Comment (ColinJPage) [inline app/public/board.js:32]: Let's delete this

**Review by ColinJPage — DISMISSED** [review]

File: app/clientsHandler.js [hunk 4]
```diff
@@ -18,10 +20,15 @@ const CLIENTS_HANDLER = (() => {
      * @param {string} gameId
      * This method registers a clientId to a specific game
      */
-    const registerClient = (clientId, username, gameId) => {
+    const registerClient = (clientId, username, gameId, token) => {
+        let loggedIn = false;
```
Comment (RichardLam772001) [inline app/clientsHandler.js:24]: This can be replaced with
```
loggedIn = token in TokenStorage;
```

File: app/game/game.js [hunk 5]
```diff
@@ -95,6 +103,41 @@ const GAME = (humanUsername, aiUsername, gameId) => {
...
+            CLIENTS_HANDLER.updatePlayerStats(HUMAN_USERNAME, 'win');
+            CLIENTS_HANDLER.updatePlayerStats(AI_USERNAME, 'win');
+            removeGame(gameId);
+        } else {
+            console.log("At least one user is not loggedin");
+            removeGame(gameId);
+        }
+    }
+
+    const onShipExplode = (gameId,AI_USERNAME,HUMAN_USERNAME) => {
```
Comment (RichardLam772001) [inline app/game/game.js:123]: We don't need AI_USERNAME, HUMAN_USERNAME as variables, these are globals and already accessible in the scope of this enclosing function. Also we have a function sendDataToBothPlayers we can use, and we probably want to send the data to players no matter what.
Suggestions:
```
    const onShipExplode = (gameId) => {
        if (CLIENTS_HANDLER.areBothPlayersLoggedIn(gameId, AI_USERNAME, HUMAN_USERNAME)) {
            CLIENTS_HANDLER.updatePlayerStats(HUMAN_USERNAME, 'lose');
            CLIENTS_HANDLER.updatePlayerStats(AI_USERNAME, 'lose');
        }
        sendDataToBothPlayers(GameEndData('lose'));
        removeGame(gameId);
    }
```

File: app/public/board.js [hunk 6]
```diff
@@ -102,6 +104,19 @@ const BOARD = ((rowCount, columnCount) => {
         roomElement.setThreat("");
     }
 
+    function showGameEndMessage(result) {
+        const messageContainer = document.getElementById("message-container");
```
Comment (RichardLam772001) [inline app/public/board.js:108]: These two variables are already defined above, so we can remove these two

**Review by RichardLam772001 — APPROVED** [review]
Ah I see what's happening, I believe the comments are from your old PR. This new PR looks perfect. I made some small comments but resolved them myself and pulled in master so you have no merge conflicts. Approved. Checked it out and works perfectly. I added a link on the game over screen to route the users back to lobby page to join another game. Very very very amazing work Meng!

**Pull request merged**